packages/demo/backend/src/services/faucet.ts

```diff
@@ -0,0 +1,149 @@
+import type { UserOperationTransactionReceipt } from 'node_modules/@eth-optimism/actions-sdk/dist/types/index.js'
+import type { Address, Hash, Hex, TypedDataDomain } from 'viem'
+import {
+  createPublicClient,
+  createWalletClient,
+  encodeFunctionData,
+  encodePacked,
+  http,
+  keccak256,
+} from 'viem'
+import { privateKeyToAccount } from 'viem/accounts'
+import { optimismSepolia } from 'viem/chains'
+
+import { faucetAbi } from '@/abis/ethFaucet.js'
+import { getActions } from '@/config/actions.js'
+import { env } from '@/config/env.js'
+
+type DripParams = {
+  recipient: Address
+  nonce: Hash
+  data: Hash
+  gasLimit?: number
+}
+
+type AuthParams = {
+  module: Address
+  id: Hash
+  proof: Hash
+}
+
+export async function isWalletEligibleForFaucet(
+  walletAddress: Address,
+): Promise<boolean> {
+  const publicClient = createPublicClient({
+    chain: optimismSepolia,
+    transport: env.OP_SEPOLIA_RPC_URL ? http(env.OP_SEPOLIA_RPC_URL) : http(),
+  })
+  const balance = await publicClient.getBalance({ address: walletAddress })
+  if (balance > 0) {
+    return false
+  }
+  // TODO: When aave lending is implemented, check if the wallet already has a position open
+  // and return false if it does.
+
+  return true
+}
+
+export async function dripEthToWallet(
+  walletAddress: Address,
+): Promise<UserOperationTransactionReceipt> {
+  const id = keccak256(walletAddress)
+  const famAddress = env.AUTH_MODULE_ADDRESS as Address
+  const domain = {
+    name: 'OffChainAuthModule',
+    version: '1',
+    chainId: optimismSepolia.id,
+    verifyingContract: famAddress,
+  }
+  const nonce = generateNonce()
+  const dripParams = createDripParams(walletAddress, nonce)
+  const authParams = await createAuthParams(
+    walletAddress,
+    famAddress,
+    id,
+    nonce,
+    domain,
+  )
+  const dripCallData = encodeFunctionData({
+    abi: faucetAbi,
+    functionName: 'drip',
+    args: [dripParams, authParams],
+  })
+  const transactionData = [
+    {
+      to: env.OP_SEPOLIA_FAUCET_ADDRESS as Address,
+      data: dripCallData,
+      value: 0n,
+    },
+  ]
+  const actions = getActions()
+  const adminSigner = privateKeyToAccount(
+    env.FAUCET_AUTH_MODULE_ADMIN_PRIVATE_KEY as Hex,
+  )
+  const adminSmartWallet = await actions.wallet.getSmartWallet({
+    signer: adminSigner,
+    deploymentSigners: [adminSigner],
+  })
+  const receipt = await adminSmartWallet.sendBatch(
+    transactionData,
+    optimismSepolia.id,
+  )
+  return receipt
+}
+
+function createDripParams(recipientAddress: Address, nonce: Hash): DripParams {
+  return {
+    recipient: recipientAddress,
+    nonce,
+    data: '0x',
+  }
+}
+
+function generateNonce(): `0x${string}` {
+  const timestamp = Date.now() + Math.random() * 1000000
+  return keccak256(encodePacked(['uint256'], [BigInt(Math.floor(timestamp))]))
+}
```

Comment:
🤔 GPT claims this gives us a ~50% chance of at least one collision after ~1,177 nonces within a 16 min window. Bro could be full of it, but suggests this as better practice regardless:

```
import { randomBytes } from 'crypto';

function generateNonce(): `0x${string}` {
  return `0x${randomBytes(32).toString('hex')}`; // 256-bit CSPRNG
}
```

Reply:
nice, that is cleaner. just updated it!